### PR TITLE
systemd-boot-builder.py: add support for profiles

### PR DESCRIPTION
**Pull request by KaiHa**
This will also list the generations of other profiles than `system` in
the boot menu. See the documentation of the `--profile-name` option of
nixos-rebuild for more information on profiles.

###### Motivation for this change

To boot other profiles with systemd-boot.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by grahamc** [comment]
What is the use case for this?

**Comment by Mic92** [comment]
Suppose you want to have different Kernel version. You can create a profile for each and get a boot menu entry to select them.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py [hunk 1]
```diff
@@ -77,36 +88,45 @@ def mkdir_p(path):
...
+    rex_profile = re.compile("^@efiSysMountPoint@/loader/entries/nixos-(.*)-generation-.*\.conf$")
+    rex_generation = re.compile("^@efiSysMountPoint@/loader/entries/nixos.*-generation-(.*)\.conf$")
     known_paths = []
     for gen in gens:
-        known_paths.append(copy_from_profile(gen, "kernel", True))
-        known_paths.append(copy_from_profile(gen, "initrd", True))
-    for path in glob.iglob("@efiSysMountPoint@/loader/entries/nixos-generation-[1-9]*.conf"):
+        known_paths.append(copy_from_profile(*gen, "kernel", True))
+        known_paths.append(copy_from_profile(*gen, "initrd", True))
+    for path in glob.iglob("@efiSysMountPoint@/loader/entries/nixos-*-generation-[1-9]*.conf"):
```
Comment (Mic92) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: Would not this fail to match old entries?

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py [hunk 2]
```diff
@@ -77,36 +88,45 @@ def mkdir_p(path):
...
         universal_newlines=True)
     gen_lines = gen_list.split('\n')
     gen_lines.pop()
-    return [ int(line.split()[0]) for line in gen_lines ]
+    return [ (profile, int(line.split()[0])) for line in gen_lines ]
 
 def remove_old_entries(gens):
-    slice_start = len("@efiSysMountPoint@/loader/entries/nixos-generation-")
-    slice_end = -1 * len(".conf")
+    rex_profile = re.compile("^@efiSysMountPoint@/loader/entries/nixos-(.*)-generation-.*\.conf$")
```
Comment (Mic92) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: Here as well. We need backward-compatible code.

**Comment by Mic92** [comment]
Currently fails to build, when I give no profile argument to `nixos-rebuild switch`:

```
created 10635 symlinks in user environment                                                                                                                                  
install-info: warning: no info dir entry in `/nix/store/1zjrj9rna9gs16d36kc53l6dzy9kkxab-system-path/share/info/time.info'                                                  
building path(s) ‘/nix/store/w9kjfxkm8hfvrfqlxxr0hilfivxl81cn-dbus-1’                                                                                                       
building path(s) ‘/nix/store/y8s1j7cv8maf7i7crjajdmvci4lg5x2v-unit-polkit.service’                                                                                          
building path(s) ‘/nix/store/jqp8b44i1lx82i4h4fpialqxl1qdssdd-unit-systemd-fsck-.service’                                                                                   
building path(s) ‘/nix/store/0p423ar4pj10jxa5277pprm0bn5paavr-unit-dbus.service’                                                                                            
building path(s) ‘/nix/store/6wx3l0v74p32f8icrq9vmkw370m7kdqw-system-units’                                                                                                 
building path(s) ‘/nix/store/h4z06sv1nys9sl6mrndk749np9d76bh2-user-units’                                                                                                   
building path(s) ‘/nix/store/gyxmp1np5zlk7vnilz71qkc3hr5rppmm-etc’                                                                                                          
building path(s) ‘/nix/store/m0w6z41fsj1qdpmw34bxcpq729qngpl8-nixos-system-turingmachine-17.09.git.5334e85’                                                                 
Traceback (most recent call last):                                                                                                                                          
  File "/nix/store/xns7s94x8a7z24hi9qd9jzrsl445sscz-systemd-boot-builder.py", line 182, in <module>                                                                         
    main()                                                                                                                                                                  
  File "/nix/store/xns7s94x8a7z24hi9qd9jzrsl445sscz-systemd-boot-builder.py", line 165, in main                                                                             
    for profile in get_profiles():                                                                                                                                          
  File "/nix/store/xns7s94x8a7z24hi9qd9jzrsl445sscz-systemd-boot-builder.py", line 127, in get_profiles                                                                     
    for x in os.listdir("/nix/var/nix/profiles/system-profiles/")                                                                                                           
FileNotFoundError: [Errno 2] No such file or directory: '/nix/var/nix/profiles/system-profiles/'                                                                            
warning: error(s) occurred while switching to the new configuration 
```

**Comment by KaiHa** [comment]
> Currently fails to build, when I give no profile argument to nixos-rebuild switch:

Ohh, I should have checked the case that no profile exists. My bad, I will update the PR but will only have time earliest on Monday to do so.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py (hunk 2 above)
Comment (KaiHa) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: This is with intent. The `rex_profile` regex is used to decide it it is a system profile or not.

File: nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py (hunk 1 above)
Comment (KaiHa) [inline nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py]: Yes, you are right I need to fix this. I will update the PR.

**Comment by KaiHa** [comment]
@Mic92, let me know if there is still something that needs to be fixed.

**Pull request merged**